### ISSUE TITLE
RFE/support request: support stop loss orders

### ISSUE THREAD

**Issue by sven337** [pmorissette/bt#81]
Hello,
I don't think it is currently possible for a user of bt to implement relative stop loss orders. These are, according to the litterature, sometimes useful for a momentum strategy, so it would be nice to have support for them.
http://www.dualmomentum.net/2015/06/momentum-and-stop-losses.html

Do you think bt is able to do this as is? Maybe by implementing a strategy that runs daily and checks for the drawdown to decide whether or not to trigger the stop loss order, on top of the monthly momentum strategy? I am not sure whether the tree structure would work.

Thank you

**Comment by pmorissette** [comment]
Hey @sven337,

You should be able to program an Algo that does this. Basically, if you had a monthly strategy `[RunMonthly(), ...]`, you would put your `CheckStopLoss` algo first in the stack to ensure it always runs. Alternatively, you could add the `run_always=True` attribute on the Algo (see `bt.algos.run_always`). 

The `CheckStopLoss` would simply need to run the stop-loss logic and call `target.close(c)` when appropriate.

I've personally never been very fond of stop losses, but if you do implement something, don't hesitate to submit a PR as I'm sure others will find use for it!

Cheers,
Phil

**Comment by sven337** [comment]
Thanks for the suggestion on how to implement this.
I'm not going to send a PR because my simulations show that stop losses are losing me money.  I will dig a little, as I believe I may have a bug, but it's likely that I'll give up on using stop loss orders. 

Below is what I implemented, in case anyone finds it useful. It supports regular & trailing stop loss orders, takes a threshold as parameter, and supports a "fallback" ticker to use instead of cash. The trailing stoploss mode needs a custom Rebalance() function that sets    ``` target.perm['stoploss_set_up_on'][item[0]] = target.now```.

Closing request: bt's design is quite sufficient for stop loss orders after all.

```python
class StopLoss(bt.Algo):
    """
    Implements a stop loss (regular or trailing). Sells assets if they go below a certain threshold under their maximum value since they were bought.

    Args:
        * threshold (float): threshold of loss that triggers sale. Use 0.05 for 5%.
        * trailing (bool): implement a trailing stop loss if true (use the highest price as reference). 
                           If false, implement a regular stop loss (use the buy price as reference).
        * target_ticker (string): name of the ticker to which to reallocate capital when stoploss orders are triggered.
                           By default, keep proceeds in cash.
    """

    def __init__(self, threshold, trailing=False, target_ticker=None):
        super(StopLoss, self).__init__()
        self.threshold = threshold
        self.trailing = trailing
        self.target_ticker = target_ticker

    def __call__(self, target):
        # For each current position, check its drawdown starting on the date it was purchased

        # If no stoploss was previously set, bail
        if target.perm.get('stoploss_set_up_on') == None:
            target.perm['stoploss_set_up_on'] = {}
            return True

        # Build list of currently held securities
        l = [ target.children[cname] for cname in target.children if target.children[cname]._weight > 0 ]
        for c in l:
            try:
                stoploss_from = target.perm['stoploss_set_up_on'][c.name]
            except KeyError:
                continue
            if not self.trailing:
                # Regular stop loss
                price_base = c.prices.ix[stoploss_from]
                loss = (c.price / price_base) - 1
                triggered = loss < -self.threshold
            else:
                # Trailing stop loss, use ffn's to_drawdown_series
                d = c.prices.ix[stoploss_from:target.now].to_drawdown_series()
                d = d[d < -self.threshold]
                triggered = (len(d) > 0)
                
            if triggered:
                print("Triggering stoploss ",  c.name, target.now, c.price, stoploss_from, c.value)
                reallocate_weight = c.weight
                target.close(c.name)
                if (self.target_ticker):
                    target.rebalance(reallocate_weight, self.target_ticker)

        return True
```

**Comment by Gillinghammer** [comment]
Feels obvious to me that a stop loss and trailing stop loss be available in this library. Wondering why people would run an algorithm without any regard to stop loss management...
